src/gf.c

```diff
@@ -2418,8 +2430,8 @@ static int ml_matches_visitor(jl_typemap_entry_t *ml, struct typemap_intersectio
             // Instead we should have computed the intersection as (UnionAll S Type{T{S}}),
             // which is a bigger type that would not have been a subtype of the prior
             // match (prior_ti). We simulate that for now by checking jl_has_typevars.
-            if (jl_is_leaf_type(prior_ti) && !jl_has_typevars(closure->match.ti) && !jl_has_typevars(prior_ti) &&
-                jl_subtype(closure->match.ti, prior_ti, 0)) {
+            if (jl_is_leaf_type(prior_ti) && !jl_has_free_typevars(closure->match.ti) && !jl_has_free_typevars(prior_ti) &&
+                jl_subtype(closure->match.ti, prior_ti)) {
```

Comment:
this conditional "feels" wrong now (although I think it is wrong in the "safe" direction of over-approximation). We should be computing the intersection in that comment above correctly now, so it seems like this just needs to be testing whether this signature is shadowed (`jl_subtype(closure->match.ti, prior_ti)`), and assume that will be correct in all cases, rather than just the leaf-type case.